tools/diag/gcollect/src/gcollect/main.go

```diff
@@ -0,0 +1,906 @@
+      }
+    }
+
+    if len(failed_nodes_array) > 0 {
+      fmt.Println("Failed nodes: ", failed_nodes_array)
+    }
+    return
+  }
+
+  // Collect logs regarding a certain request
+  if request_id != "" {
+    // e.g., RESTPP_1_1.1557527283332
+    tokens := strings.Split(request_id, ".")
+    if len(tokens) < 2 || len(tokens[1]) < 10 {
+      fmt.Printf("Illegal request_id: \"%s\"\n", request_id)
+      log.Fatal("A valid request_id shoule be like this: RESTPP_1_1.1557527283332")
+    }
+    startEpoch, err = strconv.ParseInt(tokens[1][0:10], 10, 64)
+    startEpoch = startEpoch - int64(before_seconds)
+    if err != nil || startEpoch < 0 || startEpoch > time.Now().Unix() {
+      fmt.Printf("Illegal request_id: \"%s\"\n", request_id)
+      log.Fatal("A valid request_id shoule be like this: RESTPP_1_1.1557527283332")
+    }
+    if duration > 0 {
+      endEpoch = startEpoch + int64(duration)
+    } else {
+      endEpoch = time.Now().Unix()
+    }
+
+    // append the "request_id" to the "patterns" array,
+    // so that lines that match any of them will be printed.
+    patterns = append(patterns, request_id)
+  }
+
+  if *debugPtr {
+    fmt.Println("startEpoch: ", startEpoch)
+    fmt.Println("endEpoch: ", endEpoch)
+  }
+
+  outFolder := ""
+  if slave {
+    outFolder = slave_tmp_folder + "/"
+  } else {
+    outFolder = output_dir + "/" + node_id + "/"
+    // Some debug info only need to be collected from master
+    if components == "" {
+      collectDebugInfo()
+    }
+  }
+
+  // Only collect yaml files and loader logs when components is not specified.
+  if components == "" {
+    utils.LocalRun("cp -r " + tgRootDir + "/gstore/0/part/config.yaml " + outFolder)
+    utils.LocalRun("cp -r " + tgRootDir + "/gstore/0/1/ids/graph_config.yaml " + outFolder)
+    utils.LocalRun("cp -r " + tgRootDir + "/logs/restpp/restpp_loader_logs " + outFolder)
+  }
+
+  // If there is no other regex except request_id, only gpe/gse/restpp logs will be collected.
+  if request_id != "" && len(patterns) == 1 {
+    logs.FilterLogs(getLogDir(tgRootDir, "rest", node_id),
+      outFolder + "/restpp.log", patterns, startEpoch, endEpoch, before, after)
+    logs.FilterLogs(getLogDir(tgRootDir, "gpe", node_id),
+      outFolder + "/gpe.log", patterns, startEpoch, endEpoch, before, after)
+    logs.FilterLogs(getLogDir(tgRootDir, "gse", node_id),
+      outFolder + "/gse.log", patterns, startEpoch, endEpoch, before, after)
+  } else {
+    // Otherwise, collect all the logs.
+    filterLogs(outFolder)
+  }
+
+  if !slave {
+    // Waiting for slave nodes to copy logs backup to master
+    syncSlaveNodes()
+    // Close all ssh connections to slave nodes
+    for name, client := range name2client {
+      if name == "m1" {
+        continue
+      }
+      client.Close()
+    }
+    if display {
+      printToScreen()
+    }
+  } else {
+    // Copy log files back to master on slave nodes.
+    err = utils.Scp(username, "", master_ip, sshPort,
```

Comment:
The ssh timeout was set to 30s . If any machine fails, it'll wait 30s at most, then will proceed to handle other machines. If "scp" failed for a machine, it will not try to launch the binary on that machine any more. If there's any machine failed, the nod's id will be printed at the end of the output.